runtime/lua/vim/treesitter.lua

```diff
@@ -106,6 +106,16 @@ function M.get_parser(bufnr, lang, opts)
   elseif parsers[bufnr] == nil or parsers[bufnr]:lang() ~= lang then
     assert(lang, 'lang should be valid')
     parsers[bufnr] = M._create_parser(bufnr, lang, opts)
+  elseif opts.reload then
+    opts.reload = nil -- do not pass reload to _create_parser
+    assert(lang, 'lang should be valid')
+    parsers[bufnr]:destroy()
+    parsers[bufnr] = M._create_parser(bufnr, lang, opts)
+
+    if M.highlighter.active[bufnr] then
+      M.highlighter.active[bufnr]:destroy()
+      M.highlighter.new(parsers[bufnr])
+    end
```

Comment:
why is this different than `LanguageTree:invalidate()` ? in fact we have something called `_on_reload` that currently calls invalidate(): https://github.com/neovim/neovim/blob/c1396afa7c41c2009d06a54c8aaf1ca5689a9786/runtime/lua/vim/treesitter/languagetree.lua#L1042-L1044

Would be very helpful to determine which semantics are correct and tighten things up

Reply:
`invalidate()` does not reload `LanguageTree._inject_query` which is the reason I want to reload the parser. If there is different way it should be done, I am happy to do it.

What happens there is basically what is going to happen if `:edit` executed. When `:edit` executed fires event "detach" for buffer, which triggers `detach_cb` and this deletes parser. Then `vim.treesitter.start()` called by ft plugin and creates new parser.

https://github.com/neovim/neovim/blob/4e5c633ed4871a948aff7338b793ac5f93484153/runtime/lua/vim/treesitter.lua#L47-L52

Right now my goal is to make parser read queries from filesystem again, because I change `vim.opt.rtp` where I add more queries from plugin.

Reply:
I was thinking, maybe a better idea would be refreshing queries in `invalidate`? Something like this
```lua
function LanguageTree:invalidate(reload, refresh_queries)
  self._valid = false

  if refresh_queries then
    if not self._opts.injections then
      query._reset_cache(self:lang(), 'injections')
      self._injection_query = query.get(self:lang(), 'injections')
    end
    self:_do_callback('changedquery', self:lang())
  end
    -- ...
```

and then highlighter.lua would register callback

```lua
 -- ...
    on_changedquery = function(lang)
      -- just remove queires, they are loaded later on demand
      query._reset_cache(lang, 'highlights')
      self._queries[lang] = nil
    end,
```

Reply:
`invalidate` is only supposed to be called when the tracked state of the Languagetree is not valid against the parse tree in tree-sitter, and is called fairly often. I expect clearing file system caches will kill performance.

Reply:
> `invalidate` is only supposed to be called when the tracked state of the Languagetree is not valid against the parse tree in tree-sitter, and is called fairly often

I attempted to mention this in the docstring of invalidate() https://github.com/neovim/neovim/pull/28747